rclcpp_examples/src/parameters/ros2param.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+
+#define USAGE "USAGE:\n  ros2param get <node/variable>\n  ros2param set <node/variable> <value>"
+
+typedef enum
+{
+  PARAM_NONE,
+  PARAM_GET,
+  PARAM_SET,
+} param_operation_t;
+
+rclcpp::parameter::ParameterVariant
+parse_args(int argc, char ** argv, std::string & remote_node, param_operation_t & op)
+{
+  if (argc < 3) {
+    return rclcpp::parameter::ParameterVariant();
+  }
+
+  std::string verb = argv[1];
+  std::string name = argv[2];
+  size_t slash = name.find('/');
+  if ((slash == std::string::npos) ||
+    (slash == 0) ||
+    (slash == (name.size() - 1)))
+  {
+    return rclcpp::parameter::ParameterVariant();
+  }
+  remote_node = name.substr(0, slash);
+  std::string variable = name.substr(slash + 1, name.size() - slash - 1);
+
+  if ((verb == "get") && (argc == 3)) {
+    op = PARAM_GET;
+    return rclcpp::parameter::ParameterVariant(variable, 0);
+  } else if ((verb == "set") && (argc == 4)) {
+    op = PARAM_SET;
+    std::string value = argv[3];
+    char * endptr;
+    int l = strtol(value.c_str(), &endptr, 10);
+    if ((errno == 0) && (*endptr == '\0')) {
+      return rclcpp::parameter::ParameterVariant(variable, l);
+    } else {
+      errno = 0;
+      double d = strtod(value.c_str(), &endptr);
+      if ((errno == 0) && (*endptr == '\0')) {
+        return rclcpp::parameter::ParameterVariant(variable, d);
+      } else {
+        if ((value == "true") || (value == "True")) {
+          return rclcpp::parameter::ParameterVariant(variable, true);
+        } else if ((value == "false") || (value == "False")) {
+          return rclcpp::parameter::ParameterVariant(variable, false);
+        } else {
+          return rclcpp::parameter::ParameterVariant(variable, value);
+        }
+      }
+    }
+  } else {
+    return rclcpp::parameter::ParameterVariant();
+  }
```

Comment:
To avoid the code to be nested (in this case four levels) I would suggest to utilize that the code returns in many locations. The following code doesn't need to be nested in an else block which reduces the overall nesting (and I think improves the readability since as a reader I don't have to consider all conditions of the parent blocks). For example like this (which has only two levels):

```
  if ((verb == "get") && (argc == 3)) {
    op = PARAM_GET;
    return rclcpp::parameter::ParameterVariant(variable, 0);
  }
  if ((verb == "set") && (argc == 4)) {
    op = PARAM_SET;
    std::string value = argv[3];
    char * endptr;
    int l = strtol(value.c_str(), &endptr, 10);
    if ((errno == 0) && (*endptr == '\0')) {
      return rclcpp::parameter::ParameterVariant(variable, l);
    }
    errno = 0;
    double d = strtod(value.c_str(), &endptr);
    if ((errno == 0) && (*endptr == '\0')) {
      return rclcpp::parameter::ParameterVariant(variable, d);
    }
    if ((value == "true") || (value == "True")) {
      return rclcpp::parameter::ParameterVariant(variable, true);
    }
    if ((value == "false") || (value == "False")) {
      return rclcpp::parameter::ParameterVariant(variable, false);
    }
    return rclcpp::parameter::ParameterVariant(variable, value);
  }
  return rclcpp::parameter::ParameterVariant();
```